xrpl/utils/tx_parser/utils.py

```diff
@@ -0,0 +1,1054 @@
+"""Utils for balance_changes and orderbook_changes."""
+from __future__ import annotations
+
+from datetime import datetime
+from decimal import Decimal
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union
+
+from pydash import compact, filter_, flatten, group_by, map_, map_values  # type: ignore
+from typing_extensions import Literal
+
+from xrpl.constants import XRPLException
+from xrpl.utils.xrp_conversions import drops_to_xrp
+
+"""
+###########################################
+Utils balance_parser and orderbook_parser
+###########################################
+"""
+
+
+class XRPLMetadataException(XRPLException):
+    """Exception for invalid transaction metadata."""
+
+    pass
+
+
+def is_valid_metadata(metadata: Dict[str, Any]) -> None:
+    """Check if the given metadata is valid.
+
+    Args:
+        metadata (Dict[str, Any]):
+            The transactions metadata
+
+    Raises:
+        XRPLMetadataException: If field Account is missing.
+        XRPLMetadataException: If field meta is missing.
+        XRPLMetadataException: If field AffectedNodes is missing or is empty.
+    """
+    if "transaction" in metadata:
+        if "Account" not in metadata["transaction"]:
+            raise XRPLMetadataException(
+                "Metadata incomplete: Metadata field 'Account' must be included."
+            )
+    elif "Account" not in metadata:
+        raise XRPLMetadataException(
+            "Metadata incomplete: Metadata field 'Account' must be included."
+        )
+
+    if "meta" not in metadata:
+        raise XRPLMetadataException(
+            "Metadata incomplete: Metadata field 'meta' must be included."
+        )
+
+    if "AffectedNodes" not in metadata["meta"] or not metadata["meta"]["AffectedNodes"]:
+        raise XRPLMetadataException("Metadata incomplete: No nodes provided.")
+
+
+def normalize_metadata(
+    metadata: Dict[
+        str,
+        Union[
+            str,
+            int,
+            Dict[str, str],
+            List[List[Dict[str, Union[str, int]]]],
+            Dict[
+                str,
+                List[
+                    Dict[
+                        str,
+                        Dict[
+                            str,
+                            Union[str, int, Dict[str, Union[str, int, Dict[str, str]]]],
+                        ],
+                    ],
+                ],
+            ],
+        ],
+    ],
```

Comment:
Holy crap that's one hell of a type! 

I'd recommend turning this into a class or type alias, that way you can define the type once and then use it everywhere else. 

(That would also let you turn instances of `metadata: Any` above into something like `metadata: Metadata` which is nice and readable :) )

Reply:
Here's the doc on type aliases - https://mypy.readthedocs.io/en/stable/kinds_of_types.html#type-aliases

Reply:
I thought that this would be possible but did not find the right solution to that. Thank you I will change that 🙏